src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,538 @@
+        let actual = rx.borrow().clone();
+        assert!(actual.is_none());
+
+        tokio::spawn(async move {
+            refresh_task(mock, tx).await;
+        });
+
+        rx.changed().await.unwrap();
+        let actual = rx.borrow().clone().unwrap().unwrap();
+        assert_eq!(actual, token1);
+
+        // time machine forwards time to 10 seconds before expiry
+        tokio::time::advance(Duration::from_secs(110)).await;
+
+        rx.changed().await.unwrap();
+
+        // validate that the token was refreshed within 10ish seconds
+        // before expiry
+        assert!(Instant::now() < now + Duration::from_secs(111));
+        let actual = rx.borrow().clone().unwrap().unwrap();
+        assert_eq!(actual, token2);
+    }
+
+    #[derive(Clone, Debug)]
+    struct FakeTokenProvider {
+        result: Result<Token>,
+        calls: Arc<std::sync::Mutex<i32>>,
+    }
+
+    impl FakeTokenProvider {
+        pub fn new(result: Result<Token>) -> Self {
+            FakeTokenProvider {
+                result,
+                calls: Arc::new(Mutex::new(0)),
+            }
+        }
+
+        pub fn calls(&self) -> i32 {
+            *self.calls.lock().unwrap()
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl TokenProvider for FakeTokenProvider {
+        async fn get_token(&self) -> Result<Token> {
+            // We give enough time for the a thundering herd to pile up waiting for a change notification from watch channel
+            tokio::time::sleep(Duration::from_millis(50)).await;
+
+            // Track how many calls were made to the inner token provider.
+            *self.calls.lock().unwrap() += 1;
+
+            // Return the result.
+            self.result.clone()
+        }
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 8)]
+    async fn no_initial_token_thundering_herd_success() {
+        let token = Token {
+            token: "delayed-token".to_string(),
+            token_type: "Bearer".to_string(),
+            expires_at: Some(std::time::Instant::now()),
+            metadata: None,
+        };
+
+        let tp = FakeTokenProvider::new(Ok(token.clone()));
+
+        let cache = TokenCache::new(tp.clone());
+
+        // Spawn N tasks, all asking for a token at once.
+        let tasks = (0..100)
+            .map(|_| {
+                let cache_clone = cache.clone();
+                tokio::spawn(async move { cache_clone.get_token().await })
+            })
+            .collect::<Vec<_>>();
+
+        // Wait for the N token requests to complete, verifying the returned token.
+        for task in tasks {
+            let actual = task.await.unwrap();
+            assert!(actual.is_ok(), "{}", actual.err().unwrap());
+            assert_eq!(actual.unwrap(), token);
+        }
+
+        let calls = tp.calls();
+        // Only one call to token provider should have been made
+        assert_eq!(calls, 1);
```

Comment:
This assumes that the N tasks launch within the 50ms allotted. I would not rely on timing like that.

This is what I would do, but maybe you thought the expectation was comically loose? I thought it worked.

https://github.com/dbolduc/google-cloud-rust/blob/607315bca3477f8a478fe27e8ec1d8cc1faee6bd/src/auth/src/token_cache.rs#L465-L472